core/packages/contracts/contracts/XcmApp.sol

```diff
@@ -0,0 +1,148 @@
+    event XcmExecuted(bytes32 origin, Proxy proxy, address executor, bool success);
+
+    /// @dev Called from an unauthorized sender.
+    error Unauthorized();
+
+    /// @dev The executor already exists.
+    error ExecutorExists();
+
+    /// @dev The executor was not found.
+    error ExecutorNotFound();
+
+    /// @param _registry The channel registry which is the source of messages.
+    constructor(ChannelRegistry _registry) {
+        registry = _registry;
+    }
+
+    /// @dev Approves an executor version.
+    /// @param _version An identifier for the version.
+    /// @param _executor The executor to approve.
+    function approveExecutor(bytes8 _version, address _executor) external {
+        // TODO: Should permissionless channels be able to call in here???
+        if (!registry.isInboundChannel(msg.sender)) {
+            revert Unauthorized();
+        }
+        if (executors[_version] != address(0)) {
+            revert ExecutorExists();
+        }
+        executors[_version] = _executor;
+    }
+
+    /// @dev Revokes an executor version.
+    /// @param _version An identifier for the version.
+    function revokeExecutor(bytes8 _version) external {
+        // TODO: Should permissionless channels be able to call in here???
+        if (!registry.isInboundChannel(msg.sender)) {
+            revert Unauthorized();
+        }
+        if (executors[_version] == address(0)) {
+            revert ExecutorNotFound();
+        }
+        delete executors[_version];
+    }
+
+    /// @dev Looks up the proxy and executor and executes the payload.
+    /// @param _origin The hashed origin.
+    /// @param _executor The identifier for the executor version.
+    /// @param _payload The XCM payload to be executed.
+    function dispatchToProxy(bytes32 _origin, bytes8 _executor, bytes calldata _payload) external {
+        // TODO: Should permissionless channels be able to call in here???
+        if (!registry.isInboundChannel(msg.sender)) {
+            revert Unauthorized();
+        }
+
+        Proxy proxy = proxies[_origin];
+        // JIT create proxy if it does not exist.
+        if (proxy == Proxy(address(0))) {
+            proxy = new Proxy();
+            proxies[_origin] = proxy;
+        }
+
+        // Find an approved executor.
+        address executor = executors[_executor];
+        if (executor == address(0)) {
+            revert ExecutorNotFound();
+        }
+
+        // Dispatch to proxy.
+        bool success = proxy.execute(executor, _payload);
+        emit XcmExecuted(_origin, proxy, executor, success);
+    }
+}
+
+/// @dev Executes Xcm instructions.
+contract XcmExecutor {
+    /// @dev The entry point for an payload.
+    function execute(bytes[] calldata instructions) external {
+        // TODO: registers like origin, holding, etc...
+        for (uint i = 0; i < instructions.length; i++) {
+            bytes1 instruction = instructions[i][0];
+            if (instruction == 0x00) {
```

Comment:
Lets use [Enums](https://solidity-by-example.org/enum/) to identify instructions

Reply:
Used enums and structures.